tutorials/spanruler_restaurant_reviews/scripts/preprocess_prodigy.py

```diff
@@ -0,0 +1,174 @@
+    its annotations on individual lines. Here we turn the tokens and annotations
+    into a json-format dataset and also save just the text for other functions.
+
+    Returns a dictionary with text, spans, and annotator ID and a dictionary with
+    text.
+    """
+    # open IOB data
+    with file_in.open("r", encoding="utf-8") as infile:
+        input_lines = infile.read().splitlines()
+
+    # the annotations and tokens are separated by '\t'
+    annotation_token = [line.split("\t") for line in input_lines]
+
+    text = []  # the string for each entry
+    tags = []  # the labels for each entry
+    data = []  # the full list of jsonl data with annotations
+    text_data = []  # the list of jsonl data without annotations
+
+    for lines in annotation_token:
+        if len(lines) == 2:  # contains both annotation and token
+            annotation, token = lines
+            text.append(token)
+            tags.append(annotation)
+        if len(lines) == 1:  # contains space
+            doc = Doc(Vocab(), words=text)
+            tags = iob_to_biluo(tags)
+            offsets = biluo_tags_to_offsets(doc, tags)  # (0, 7, Rating)
+            # create ents from offsets
+            ents = [
+                {"start": span[0], "end": span[1], "label": span[2]} for span in offsets
+            ]
+            # append data to lists
+            data.append(
+                {
+                    "text": " ".join(text),
+                    "spans": ents,
+                    "_annotator_id": "original_annotations",
+                    "_session_id": "original_annotations",
+                }
+            )
+            text_data.append({"text": " ".join(text)})
+            # clear lists
+            text = []
+            tags = []
+
+    return data, text_data
+
+
+def get_model_data(
+    text_data=Arg(..., help="Dictionary of texts in the dataset."),
+    trained_model=Arg(..., help="The trained NER model."),
+):
+    """
+    Create JSON data with model annotations from the trained NER model.
+
+    Returns a dictionary with text, spans, and annotator ID.
+    """
+    # load trained model
+    ner = spacy.load(trained_model)
+
+    data = copy.deepcopy(text_data)
+    for line in data:
+        text = line["text"]
+
+        doc = ner(text)
+        spans = []
+        for ent in doc.ents:
+            # Create a span dict for the predicted entity.
+            spans.append(
+                {"start": ent.start_char, "end": ent.end_char, "label": ent.label_,}
+            )
+        line["spans"] = spans
+        line["_annotator_id"] = "ner_model"
+        line["_session_id"] = "ner_model"
+
+    return data
+
+
+def get_ruler_data(text_data=Arg(..., help="Dictionary of texts in the dataset."),):
+    """
+    Create JSON data with annotations from the SpanRuler patterns.
+
+    Returns a dictionary with text, spans, and annotator ID.
+    """
+    nlp = spacy.blank("en")  # blank tokenizer
```

Comment:
The comment might not be necessary

```suggestion
    nlp = spacy.blank("en") 
```